File: Nori/Panel/ActionGrammar.swift
import Foundation

/// The single source of truth for what a key, click or number does.
///
/// Enter pastes. Modifiers are orthogonal bits: ⇧ = plain text, ⌥ = keep the panel open,
/// ⌘ = copy only. No preference can change this; the hint bar renders from the same function.
enum ActionGrammar {
    enum Base: Equatable, Sendable {
        case returnKey
        case click
        case number(Int)
    }

    struct Bits: OptionSet, Sendable, Hashable {
        let rawValue: Int
        static let plain = Bits(rawValue: 1 << 0)     // ⇧
        static let keepOpen = Bits(rawValue: 1 << 1)  // ⌥
        static let copyOnly = Bits(rawValue: 1 << 2)  // ⌘
        static let all: [Bits] = (0..<8).map { Bits(rawValue: $0) }
    }

    struct Capabilities: Sendable, Equatable {
        var accessibilityTrusted: Bool
    }

    enum Action: Equatable, Sendable {
        case paste(plain: Bool, keepOpen: Bool)
        case copy(plain: Bool, keepOpen: Bool)

        var isPaste: Bool { if case .paste = self { true } else { false } }
        var plain: Bool {
            switch self {
            case let .paste(plain, _), let .copy(plain, _): plain
            }
        }
        var keepOpen: Bool {
            switch self {
            case let .paste(_, keepOpen), let .copy(_, keepOpen): keepOpen
            }
        }
    }

    static func resolve(_ base: Base, _ bits: Bits, _ caps: Capabilities) -> Action {
        let plain = bits.contains(.plain)
        let keepOpen = bits.contains(.keepOpen)
        // ⌘ is the trigger of the number row, so it cannot also mean "copy only" there.
        let copyOnly: Bool = if case .number = base { false } else { bits.contains(.copyOnly) }
        if copyOnly || !caps.accessibilityTrusted {
            return .copy(plain: plain, keepOpen: keepOpen)
        }
        return .paste(plain: plain, keepOpen: keepOpen)
    }

    /// Verb for hint bars and menus.
    static func verb(for action: Action) -> String {
        switch action {
        case .paste(false, false): "Paste"
        case .paste(true, false): "Paste as plain text"
        case .paste(false, true): "Paste and keep Nori open"
        case .paste(true, true): "Paste plain, keep open"
        case .copy(false, false): "Copy"
        case .copy(true, false): "Copy as plain text"
        case .copy(false, true): "Copy, keep open"
        case .copy(true, true): "Copy plain, keep open"
        }
    }
}

#if canImport(AppKit)
import AppKit

extension ActionGrammar.Bits {
    /// Translate event modifiers into grammar bits.
    init(modifierFlags: NSEvent.ModifierFlags) {
        var bits: ActionGrammar.Bits = []
        if modifierFlags.contains(.shift) { bits.insert(.plain) }
        if modifierFlags.contains(.option) { bits.insert(.keepOpen) }
        if modifierFlags.contains(.command) { bits.insert(.copyOnly) }
        self = bits
    }

    /// Glyphs in the macOS order (⌃⌥⇧⌘) for keycaps.
    var glyphs: String {
        var result = ""
        if contains(.keepOpen) { result += "⌥" }
        if contains(.plain) { result += "⇧" }
        if contains(.copyOnly) { result += "⌘" }
        return result
    }
}
#endif

File: windows/Nori.Core/ActionGrammar.cs
namespace Nori.Core;

/// <summary>
/// The single source of truth for what a key, click or number does.
/// Enter pastes. Modifiers are orthogonal bits: Shift = plain text, Alt = keep the panel open,
/// Ctrl = copy only. No preference can change this; the hint bar renders from the same function.
/// </summary>
public static class ActionGrammar
{
    public abstract record Base
    {
        public sealed record ReturnKey : Base;
        public sealed record Click : Base;
        public sealed record Number(int Value) : Base;

        public static readonly Base Return = new ReturnKey();
        public static readonly Base Mouse = new Click();
    }

    [Flags]
    public enum Bits
    {
        None = 0,
        Plain = 1 << 0,     // Shift
        KeepOpen = 1 << 1,  // Alt
        CopyOnly = 1 << 2,  // Ctrl
    }

    /// <summary>Every combination of the three bits, for truth-table tests.</summary>
    public static readonly IReadOnlyList<Bits> AllBits = Enumerable.Range(0, 8).Select(i => (Bits)i).ToList();

    /// <summary>
    /// On Windows a synthetic Ctrl+V needs no special permission, so pasting is normally possible.
    /// The flag stays so a degraded environment (no foreground window to paste into) can turn Enter into Copy.
    /// </summary>
    public readonly record struct Capabilities(bool CanPaste);

    public readonly record struct Action(bool IsPaste, bool Plain, bool KeepOpen)
    {
        public static Action Paste(bool plain, bool keepOpen) => new(true, plain, keepOpen);
        public static Action Copy(bool plain, bool keepOpen) => new(false, plain, keepOpen);
        public bool IsCopy => !IsPaste;
    }

    public static Action Resolve(Base @base, Bits bits, Capabilities caps)
    {
        var plain = bits.HasFlag(Bits.Plain);
        var keepOpen = bits.HasFlag(Bits.KeepOpen);
        // Ctrl is the trigger of the number row, so it cannot also mean "copy only" there.
        var copyOnly = @base is not Base.Number && bits.HasFlag(Bits.CopyOnly);
        if (copyOnly || !caps.CanPaste)
        {
            return Action.Copy(plain, keepOpen);
        }
        return Action.Paste(plain, keepOpen);
    }

    /// <summary>Resource key of the verb for hint bars and menus (English text lives in the app's resources).</summary>
    public static string VerbKey(Action action) => (action.IsPaste, action.Plain, action.KeepOpen) switch
    {
        (true, false, false) => "Verb_Paste",
        (true, true, false) => "Verb_PastePlain",
        (true, false, true) => "Verb_PasteKeepOpen",
        (true, true, true) => "Verb_PastePlainKeepOpen",
        (false, false, false) => "Verb_Copy",
        (false, true, false) => "Verb_CopyPlain",
        (false, false, true) => "Verb_CopyKeepOpen",
        (false, true, true) => "Verb_CopyPlainKeepOpen",
    };

    /// <summary>Default English verb, used by tests and as the fallback when a resource is missing.</summary>
    public static string EnglishVerb(Action action) => (action.IsPaste, action.Plain, action.KeepOpen) switch
    {
        (true, false, false) => "Paste",
        (true, true, false) => "Paste as plain text",
        (true, false, true) => "Paste and keep Nori open",
        (true, true, true) => "Paste plain, keep open",
        (false, false, false) => "Copy",
        (false, true, false) => "Copy as plain text",
        (false, false, true) => "Copy, keep open",
        (false, true, true) => "Copy plain, keep open",
    };

    /// <summary>Modifier prefix in the Windows order (Ctrl, Alt, Shift) for keycaps, e.g. "Ctrl+Shift+".</summary>
    public static string ModifierPrefix(Bits bits)
    {
        var result = string.Empty;
        if (bits.HasFlag(Bits.CopyOnly)) result += "Ctrl+";
        if (bits.HasFlag(Bits.KeepOpen)) result += "Alt+";
        if (bits.HasFlag(Bits.Plain)) result += "Shift+";
        return result;
    }

    /// <summary>Translate a modifier state into grammar bits.</summary>
    public static Bits FromModifiers(bool shift, bool alt, bool control)
    {
        var bits = Bits.None;
        if (shift) bits |= Bits.Plain;
        if (alt) bits |= Bits.KeepOpen;
        if (control) bits |= Bits.CopyOnly;
        return bits;
    }
}
